cmd/zdb/zdb.c

```diff
@@ -9458,17 +9532,22 @@ main(int argc, char **argv)
 			free(checkpoint_target);
 	}
 
+fini:
 	if (os != NULL) {
 		close_objset(os, FTAG);
-	} else {
+	} else if (spa != NULL) {
 		spa_close(spa, FTAG);
 	}
 
 	fuid_table_destroy();
 
 	dump_debug_buffer();
 
-	kernel_fini();
+	if (g_zfs)
+		libzfs_fini(g_zfs);
+
+	if (kernel_init_done)
+		kernel_fini();
```

Comment:
> When encountering an error or early return, kernel_fini is not called.

Can you split the `kernel_fini()` in to its own patch and open a new PR for it.  There's no reason this needs to be blocked by the larger cache file detection improvements.

Reply:
Thanks. Opened https://github.com/openzfs/zfs/pull/16152.